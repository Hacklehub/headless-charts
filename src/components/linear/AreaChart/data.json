[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
