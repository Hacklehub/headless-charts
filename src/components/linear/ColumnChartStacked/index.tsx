import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { max, sum } from 'd3-array';
import { pointer, select, selectAll } from 'd3-selection';
import { scaleBand, scaleLinear } from 'd3-scale';
/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useEffect } from 'react';

import { format } from 'd3-format';
import { transition } from 'd3';

interface DataItem {
  [key: string]: number | string;
}

interface ColumnChartStackedProps {
  data: DataItem[];
  id: string;
  className?: string;
  x: {
    key: string;
    axis?: 'top' | 'bottom';
  };
  y: {
    key: string;
    axis?: 'left' | 'right';
    className: string;
  }[];
  margin?: {
    left: number;
    right: number;
    top: number;
    bottom: number;
  };
  padding?: {
    left: number;
    right: number;
    bottom: number;
    top: number;
  };
  paddingBar?: number;
  waterfall?: boolean;
  tooltip?: {
    className?: string;
    html?: (data: any) => string;
    keys?: string[];
  };
  referenceLines?: {
    y: number;
    className?: string;
  }[];
  tickFormat?: string;
  drawing?: {
    duration?: number;
  };
}

interface drawHLineProps {
  x: number;
  y: number;
  direction?: 'left' | 'right';
  className?: string;
  dashed?: boolean;
}

const ColumnChartStacked = ({
  data = [],
  id,
  className,
  x,
  y,
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 40,
  },
  padding = {
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
  },
  paddingBar = 0.2,
  waterfall,
  tooltip,
  referenceLines = [],
  tickFormat,
  drawing = undefined,
}: ColumnChartStackedProps) => {
  const formatMapping = {
    '%': '.0%',
    $: '($.2f',
    tri: ',.2r',
    hex: '#x',
    SI: '.2s',
  };

  const refreshChart = useCallback(() => {
    /* eslint-disable */
    const svg = select(`#${id}`);
    svg.selectAll('*').remove();
    const g = svg.append('g');

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const xFn = scaleBand()
      // @ts-ignore
      .domain(data.map((d) => d[x.key]))
      .range([margin.left + padding.left, width - margin.right - padding.right])
      .padding(paddingBar);

    const yFnRange = [
      height - margin.bottom - padding.bottom,
      margin.top + padding.top,
    ];

    const yFn = scaleLinear()
      // @ts-ignore
      .domain([0, max(data.map((d) => sum(y.map((value) => d[value.key]))))])
      .range(yFnRange);

    y.map((column, i) => {
      const barsG = g.append('g');
      const beforeColumns = y.filter((_, idx) => idx <= i).map((c) => c.key);

      const bars = barsG
        .selectAll('g')
        .data(data)
        .enter()
        .append('rect')
        .attr('class', `${column.className} fill-current`)
        .attr('z-index', 100 - i)
        .attr(
          'x',
          (d) =>
            // @ts-ignore
            xFn(d[x.key]) + (waterfall ? (xFn.bandwidth() / y.length) * i : 0)
        )
        .attr('y', (d: any) => yFn(sum(beforeColumns.map((c: any) => d[c]))))
        .style('z-index', 10 + i)
        .attr('width', () =>
          waterfall
            ? // @ts-ignore
              xFn.bandwidth() / y.length - (waterfall.padding || 0)
            : xFn.bandwidth()
        )
        .attr('height', (d: any) =>
          drawing?.duration ? 0 : yFn(0) - yFn(d[column.key])
        )
        .on('mouseenter', function (event, d) {
          if (tooltip) {
            tooltipDiv.style('opacity', 1);
            const [bX, bY] = pointer(event, select('body'));
            tooltipDiv
              .style('left', `${bX + 10}px`)
              .style('top', `${bY + 10}px`);
            tooltipDiv.html(
              tooltip && tooltip.html
                ? tooltip.html(d)
                : tooltip.keys
                ? tooltip.keys
                    .map((key) => `${key}: ${d[key] || ''}`)
                    .join('<br/>')
                : `${d[x.key]} <br/> ${column.key} ${
                    tickFormat
                      ? // @ts-ignore
                        formatMapping[tickFormat]
                        ? // @ts-ignore
                          format(formatMapping[tickFormat])(d[column.key])
                        : format(tickFormat)
                      : d[column.key]
                  }`
            );
          }
        })
        .on('mousemove', function (event) {
          const [bX, bY] = pointer(event, select('body'));
          tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
        })
        .on('mouseleave', function () {
          tooltip &&
            tooltipDiv
              .style('opacity', '0')
              .style('left', `0px`)
              .style('top', `0px`);
        });

      transition();

      if (drawing?.duration) {
        bars
          .transition()
          .duration(drawing.duration)
          .attr('height', (d: any) => yFn(0) - yFn(d[column.key]));
      }
    });

    function drawHLine({
      x,
      y,
      direction = 'left',
      className,
      dashed = false,
    }: drawHLineProps) {
      const horizontalLine = g
        .append('line')
        .attr('class', mergeTailwindClasses(className, 'line stroke-current'))
        .attr('x1', direction === 'left' ? margin.left : x)
        .attr('x2', direction === 'left' ? x : width + margin.left)
        .attr('y1', y)
        .attr('y2', y)
        .attr('clip-path', 'url(#clip)')
        .attr('stroke', '#dddddd');
      dashed && horizontalLine.style('stroke-dasharray', '10,5');
    }

    referenceLines.map((object) => {
      object.y &&
        drawHLine({
          x: width - margin.right,
          y: yFn(object.y),
          className: `${object.className || ''} reference-line`,
        });
    });

    const tooltipDiv = select('body')
      .append('div')
      .attr('id', 'tooltip')
      .style('position', 'absolute')
      .style('opacity', '0')
      .attr('class', `tooltip ${(tooltip && tooltip.className) || ''}`);

    // @ts-ignore
    const yAxis = y.axis === 'right' ? axisRight(yFn) : axisLeft(yFn);

    tickFormat &&
      yAxis.tickFormat(
        // @ts-ignore
        formatMapping[tickFormat]
          ? // @ts-ignore
            format(formatMapping[tickFormat])
          : format(tickFormat)
      );

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        // @ts-ignore
        `translate(${y.axis === 'right' ? margin.left + width : margin.left},0)`
      );
    yAxisG.call(yAxis);

    const xAxis = x.axis === 'top' ? axisTop(xFn) : axisBottom(xFn);

    const xAxisG = g.append('g').attr('class', 'axis--x axis');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top : height - margin.bottom
        })`
      )
      .call(xAxis);
  }, [data]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data]);
  /* eslint-enable */
  return (
    <svg
      id={id}
      className={mergeTailwindClasses(defaultChartClassNames, className)}
    />
  );
};

export default ColumnChartStacked;
