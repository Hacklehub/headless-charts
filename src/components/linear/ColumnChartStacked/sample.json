[
  { "name": "A", "value1": 30, "value2": 20, "value3": 10 },
  { "name": "B", "value1": 40, "value2": 25, "value3": 15 },
  { "name": "C", "value1": 20, "value2": 35, "value3": 5 }
]
