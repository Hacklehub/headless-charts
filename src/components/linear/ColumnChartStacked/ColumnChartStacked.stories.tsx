/* eslint-disable @typescript-eslint/ban-ts-comment */

import ColumnChartStacked from '.';
import { Meta } from '@storybook/react';
import data from './sample.json';

const meta: Meta<typeof ColumnChartStacked> = {
  title: 'Linear/ColumnChartStacked',
  component: ColumnChartStacked,
  tags: ['autodocs'],
};

export default meta;

export const Default = {
  args: {
    data,
    id: 'column-chart-stack-default',
    y: [
      {
        key: 'macbook',
      },
      {
        key: 'iphone',
      },
      {
        key: 'ipad',
      },
      {
        key: 'wearables',
      },
      {
        key: 'services',
      },
    ],
    x: { key: 'year' },
  },
};

export const Styled = {
  args: {
    ...Default.args,
    id: 'column-chart-stack-styled',
    className: 'bg-gray-100 rounded',
    padding: {
      top: 20,
      right: 20,
      bottom: 20,
      left: 5,
      bar: 0.1,
    },
    margin: {
      top: 10,
      right: 40,
      bottom: 40,
      left: 60,
    },
    y: [
      {
        key: 'macbook',
        className: 'fill-purple-800',
      },
      {
        key: 'iphone',
        className: 'fill-purple-600',
      },
      {
        key: 'ipad',
        className: 'fill-purple-400',
      },
      {
        key: 'wearables',
        className: 'fill-purple-300',
      },
      {
        key: 'services',
        className: 'fill-purple-200',
      },
    ],
  },
};

export const WithTooltip = {
  args: {
    ...Styled.args,
    tooltip: {
      className: 'bg-gray-100 rounded p-2',
    },
  },
};

export const WithCustomTooltip = {
  args: {
    ...Styled.args,
    tooltip: {
      html: (data: any) => {
        return `
          <div class="bg-gray-100 rounded p-2">
            <div class="text-sm font-semibold">${data.name}</div>
            <div class="text-xs">${data['value1']}</div>
          </div>
        `;
      },
    },
  },
};

export const WithDrawing = {
  args: {
    ...WithTooltip.args,
    drawing: {
      duration: 1000,
    },
  },
};
export const Waterfall = {
  args: {
    ...Styled.args,
    id: 'column-chart-stack-waterfall',
    waterfall: true,
  },
};

export const WaterfallDrawing = {
  args: {
    ...Waterfall.args,
    drawing: {
      duration: 1000,
    },
  },
};
