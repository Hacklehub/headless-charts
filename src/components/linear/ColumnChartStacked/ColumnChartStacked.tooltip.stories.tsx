/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-explicit-any */

import { Meta, StoryObj } from '@storybook/react';

import ColumnChartStacked from '.';
import data from './sample.json';

const meta: Meta<typeof ColumnChartStacked> = {
  title: 'Linear/ColumnChartStacked/Tooltips',
  component: ColumnChartStacked,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof ColumnChartStacked>;

export const Tooltip: Story = {
  args: {
    id: 'tooltip',
    data,
    x: { key: 'name' },
    padding: {
      top: 20,
      right: 20,
      bottom: 20,
      left: 5,
    },
    paddingBar: 0.1,
    margin: {
      top: 10,
      right: 40,
      bottom: 40,
      left: 60,
    },
    y: [
      { key: 'value1', className: 'text-purple-500' },
      { key: 'value2', className: 'text-purple-700' },
      { key: 'value3', className: 'text-purple-900' },
    ],
    className: 'bg-gray-100 rounded',
    tooltip: {
      className: 'bg-gray-100 rounded p-2',
    },
  },
};

export const TooltipCustomHtml: Story = {
  args: {
    ...Tooltip.args,
    id: 'tooltip-custom-html',
    tooltip: {
      html: (data: any) => {
        return `
          <div class="bg-gray-100 rounded p-2">
            <div class="text-sm font-semibold">${data.name}</div>
            <div class="text-xs">${data['value1']}</div>
          </div>
        `;
      },
    },
  },
};
