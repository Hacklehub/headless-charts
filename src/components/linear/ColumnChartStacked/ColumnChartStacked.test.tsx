import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import ColumnChartStacked from '.';
import data from './sample.json';

describe('columnChartStacked', () => {
  it('renders columnChartStacked component for snapshot test', () => {
    render(
      <ColumnChartStacked
        id='columnChartStacked'
        data={data}
        x={{ key: 'name' }}
        y={[
          { key: 'value1', className: '' },
          { key: 'value2', className: '' },
          { key: 'value3', className: '' },
        ]}
      />
    );
    const columnChartStacked = screen.getByTestId('columnChartStacked');
    expect(columnChartStacked).toMatchSnapshot();
  });
  it('renders speedometer component', () => {
    render(
      <ColumnChartStacked
        id='columnChartStacked'
        data={data}
        x={{ key: 'name' }}
        y={[
          { key: 'value1', className: '' },
          { key: 'value2', className: '' },
          { key: 'value3', className: '' },
        ]}
      />
    );
    const columnChartStacked = screen.getByTestId('columnChartStacked');
    expect(columnChartStacked).toBeInTheDocument();
  });
});
