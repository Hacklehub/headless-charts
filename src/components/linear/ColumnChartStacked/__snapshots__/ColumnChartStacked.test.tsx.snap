// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`columnChartStacked > renders columnChartStacked component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="columnChartStacked"
  id="columnChartStacked"
>
  <g>
    <g>
      <rect
        class=" fill-current"
        height="-24"
        style="z-index: 10;"
        width="20.000000000000004"
        x="15.000000000000007"
        y="-16"
        z-index="100"
      />
      <rect
        class=" fill-current"
        height="-32"
        style="z-index: 10;"
        width="20.000000000000004"
        x="-9.999999999999996"
        y="-8"
        z-index="100"
      />
      <rect
        class=" fill-current"
        height="-16"
        style="z-index: 10;"
        width="20.000000000000004"
        x="-35"
        y="-24"
        z-index="100"
      />
    </g>
    <g>
      <rect
        class=" fill-current"
        height="-16"
        style="z-index: 11;"
        width="20.000000000000004"
        x="15.000000000000007"
        y="0"
        z-index="99"
      />
      <rect
        class=" fill-current"
        height="-20"
        style="z-index: 11;"
        width="20.000000000000004"
        x="-9.999999999999996"
        y="12"
        z-index="99"
      />
      <rect
        class=" fill-current"
        height="-28"
        style="z-index: 11;"
        width="20.000000000000004"
        x="-35"
        y="4"
        z-index="99"
      />
    </g>
    <g>
      <rect
        class=" fill-current"
        height="-8"
        style="z-index: 12;"
        width="20.000000000000004"
        x="15.000000000000007"
        y="8"
        z-index="98"
      />
      <rect
        class=" fill-current"
        height="-12"
        style="z-index: 12;"
        width="20.000000000000004"
        x="-9.999999999999996"
        y="24"
        z-index="98"
      />
      <rect
        class=" fill-current"
        height="-36"
        style="z-index: 12;"
        width="20.000000000000004"
        x="-35"
        y="40"
        z-index="98"
      />
    </g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,-39.5H0.5V40.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-39.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-31.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          10
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-23.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-15.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          30
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-7.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,0.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          50
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,8.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,16.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          70
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,24.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,32.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          90
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,40.5)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          100
        </text>
      </g>
    </g>
    <g
      class="axis--x axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H-39.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(25.000000000000007,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          A
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(5.329070518200751e-15,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-25,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          C
        </text>
      </g>
    </g>
  </g>
</svg>
`;
